Feature/ld2410 module device binding

MrCSharp22 opened the pull request:
<!--- In the TITLE (↑↑↑↑ above ↑↑↑↑ **NOT HERE**) provide a general, short summary of your changes -->
<!--- Please DO NOT use references to other PR's or issues -->

## Description
<!--- Describe your changes in detail -->
<!--- Bulleted list. Full sentences. Ending with a dot. -->

- Add support for the HLK-LD2410 24Ghz Human Presence Body Induction Distance Detection Radar Sensor Module. [Device info here](https://hlktech.net/index.php?id=988).

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If this **fixes** OR **closes** OR  **resolves** an open issue, please link to the issue there using the template bellow (mind the pattern to link there as all issues are tracked in the Home repository) -->
<!--- **JUST** replace NNNNN with the issue number -->
- LD2410 is a great presence sensing device that is more accurate than traditional PIR motion sensors. It uses 24Ghz mmWave and so it won't be affected by sunlight and other environmental factors.
- I need to communicate with this device using nanoFramework so I started wokring on the device binding. I'd also like nanoFramework to have official support for it so I am contributing my work.

## How Has This Been Tested?<!-- (IF APPLICABLE) -->
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Tested locally using my own LD2410 module and the official [HLKRadar-Tool](https://play.google.com/store/apps/details?id=com.hlk.hlkradartool) app to verify changes and measurements.

## Screenshots<!-- (IF APPROPRIATE): -->

- WIP

## Types of changes
<!--- What types of changes does this PR introduce? Put an `x` in all the boxes that apply: -->
- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [X] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- PLEASE PLEASE PLEASE don't tick all of them just because -->
- [X] My code follows the code style of this project (only if there are changes in source code).
- [X] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [X] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [X] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

Ellerbach commented on devices/Ld2410/Ld2410/Commands/ReadConfigurationsCommand.cs:
```diff
@@ -0,0 +1,46 @@
+        internal byte MaxMovingDistanceGate { get; set; }
+        internal byte MaxStaticDistanceGate { get; set; }
+
+        internal byte Gate0MotionSensitivity { get; set; }
```
maybe an array instead of bytes?

Ellerbach commented on devices/Ld2410/Ld2410/Commands/ReadConfigurationsCommand.cs:
```diff
@@ -0,0 +1,46 @@
+        internal byte Gate7MotionSensitivity { get; set; }
+        internal byte Gate8MotionSensitivity { get; set; }
+
+        internal byte Gate0RestSensitivity { get; set; }
```
maybe an array instead of bytes?

Ellerbach commented on devices/Ld2410/Ld2410/Commands/EnableConfigurationCommand.cs:
```diff
@@ -0,0 +1,27 @@
+﻿namespace Ld2410.Commands
+{
+    internal sealed class EnableConfigurationCommand : CommandFrame
```
in theory, it's 1 class per file, the linter will most likely complain about it.

MrCSharp22 replied:
Apologies as my StyleCop was not working. It is fixed now so I have corrected all style issues.

Ellerbach commented on devices/Ld2410/Ld2410/Extensions/EndiannessExtensions.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Ld2410.Extensions
+{
+    internal static class EndiannessExtensions
```
There is the https://github.com/nanoframework/nanoFramework.IoT.Device/tree/develop/devices/System.Buffers.Binary.BinaryPrimitives is converting anything to Little Endian. It may be better than looking at the BitConverter

MrCSharp22 replied:
Ahh very cool. I didn't know about this class. Will try to migrate as much as possible to using it. Thanks :)

Ellerbach commented on devices/Ld2410/Ld2410/Ld2410.cs:
```diff
@@ -0,0 +1,379 @@
+		/// Initializing the <see cref="Radar"/> is not enough to connect to the radar. You must call <see cref="Radar.Connect"/>
+		/// when ready to start reading measurements and controlling the radar.
+		/// </remarks>
+		public Radar(string serialPortName, int baudRate = 256_000, TimeSpan commandTimeout = default)
```
It's a good practice to add also a constructor providing already a SerialPort

Ellerbach commented on devices/Ld2410/Ld2410/Ld2410.cs:
```diff
@@ -0,0 +1,379 @@
+				throw new ArgumentOutOfRangeException();
+			}
+
+			this.Configuration = new()
```
coding wise, you don't need all the this.
We're not using it except if mandatory.

MrCSharp22 replied:
Yup, all gone now :)

Ellerbach commented on devices/Ld2410/Ld2410/Ld2410.cs:
```diff
@@ -0,0 +1,379 @@
+	/// </summary>
+	public sealed class Radar : IDisposable
+	{
+		private readonly SerialPort serialPort;
```
All the private properties should start with an underscore `_`
That's what we are adopting

Ellerbach commented on devices/Ld2410/Ld2410/Extensions/EndiannessExtensions.cs:
```diff
@@ -0,0 +1,30 @@
+        internal static byte[] EnsureLittleEndian(this byte[] valueBytes)
+        {
+            if (!BitConverter.IsLittleEndian)
+                return valueBytes.Reverse();
```
the linter should catch this :-) We're using explicit brakets.

Ellerbach commented on devices/Ld2410/Ld2410/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,17 @@
+
+/////////////////////////////////////////////////////////////////
+// This attribute is mandatory when building Interop libraries //
+[assembly: AssemblyNativeVersion("0.0.0.0")]
```
there is no need of native version, so this can go away

Ellerbach commented on devices/Ld2410/Ld2410/Reporting/EngineeringModeReportFrame.cs:
```diff
@@ -0,0 +1,122 @@
+		/// Gets the movement energy at distance gate 3.
+		/// </summary>
+		public byte Gate3MovementEnergy { get; internal set; }
+
```
The linter should catch this extra line

Ellerbach commented on devices/Ld2410/Ld2410/Reporting/EngineeringModeReportFrame.cs:
```diff
@@ -0,0 +1,122 @@
+		/// <summary>
+		/// Gets the movement energy at distance gate 0.
+		/// </summary>
+		public byte Gate0MovementEnergy { get; internal set; }
```
I really do feel a table will be better. If things goes in pair, then maybe even a class and an array of it.

MrCSharp22 replied:
This is done.

Ellerbach commented on devices/Ld2410/Ld2410/Reporting/ReportFrameParser.cs:
```diff
@@ -0,0 +1,123 @@
+            switch (data[(index += 2)])
+            {
+                case (byte)ReportingType.BasicMode:
+                    {
```
in theory, you don't need the brackets for the case

Ellerbach reviewed (commented):
Great start! Couple of comments, mainly some style. Great work!

MrCSharp22 commented:
@Ellerbach Thanks for your initial round of reviews. I have fixed the linter issues and went through your comments and made the appropriate updates. Please have another look when you can and let me know if you have anymore feedback.

Additionally, I have added a sample app and appropriate README file.

As always, thanks for taking the time.

Ellerbach commented on devices/Ld2410/Samples/Ld2410Sample/Program.cs:
```diff
@@ -0,0 +1,69 @@
+using System;
```
missing the header

Ellerbach commented on devices/Ld2410/Samples/Ld2410Sample/packages.config:
you need as well a package lock

MrCSharp22 replied:
Weird that I did not have one. I had to run `nuget restore` through the terminal manually to get one made. This is done now.

Ellerbach commented on devices/Ld2410/Samples/Ld2410Sample/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,33 @@
+﻿using System.Reflection;
```
header is missing here as well

Ellerbach reviewed (commented):
Few more comments, you're almost done! Great great work!

MrCSharp22 commented on devices/Ld2410/Ld2410.nfproj:
```diff
@@ -0,0 +1,107 @@
+    <Compile Include="BaudRate.cs" />
+    <Compile Include="Commands\SetSerialPortBaudRateCommandAck.cs" />
+    <Compile Include="GateConfiguration.cs">
+      <ExcludeFromStyleCop>True</ExcludeFromStyleCop>
```
It was complaining about the following if statement for using the new pattern matching syntax:

```csharp
if (value is < 0 or > 100) // this line
{
    throw new ArgumentOutOfRangeException();
}
```


When I looked around for solutions, it was either this or change the code. I would've changed the code except that in `GateEnergy` class, StyleCop did not like `public readonly struct GateEnergy` (readonly structs) so I opted in to just tell it to ignore those 2 files. Sadly, I don't think I came across a way to use pragmas or some C# attribute to exclude specific lines.

Ellerbach replied:
all good. It's ok, there are limitations with the linter, so it's ok not to use in some circumstances. Once we'll be in SDK, all this should be easier as well. In the mean time, it helps keeping wuite a lot of standards.

MrCSharp22 commented on devices/Ld2410/Ld2410.nfproj:
```diff
@@ -0,0 +1,107 @@
+    <TargetFrameworkVersion>v1.0</TargetFrameworkVersion>
+    <DocumentationFile>bin\$(Configuration)\Iot.Device.Ld2410.xml</DocumentationFile>
+    <StyleCopTreatErrorsAsWarnings>false</StyleCopTreatErrorsAsWarnings>
+  </PropertyGroup>
```
Yes, I forgot about code signing. Thanks for the heads up. This is done too.

MrCSharp22 commented on devices/Ld2410/Samples/Ld2410Sample/Program.cs:
```diff
@@ -0,0 +1,69 @@
+#endif
+
+#if ESP32
+			Configuration.SetPinFunction(32, DeviceFunction.COM2_RX);
```
Tabs vs Spaces issue here. Those 2 lines are using tabs for some reason.

MrCSharp22 commented on devices/Ld2410/Samples/Ld2410Sample/Ld2410Sample.nfproj:
```diff
@@ -0,0 +1,56 @@
+    <Compile Include="Properties\AssemblyInfo.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <Reference Include="mscorlib">
```
I am not sure I understand what is missing here. I added the packages that the sample app uses so I assumed that would be enough. Can you please clarify? Thanks :)

MrCSharp22 replied:
Please ignore. Seems like `Update-Nuget -Restore` fixed some of this and I had to correct some version numbers. This is fixed now and the build is passing.

MrCSharp22 commented:
Thanks for the help @Ellerbach this should be good now. If not, let me know ;)

Ellerbach commented on devices/Ld2410/GateConfiguration.cs:
```diff
@@ -0,0 +1,101 @@
+        /// <summary>
+        /// The distance, in centimeters, per each radar distance gate.
+        /// </summary>
+        public readonly static byte DistancePerGateInCm = 75;
```
can we get a Length instead of the byte? not everyone is fluent in normal units ;-)

Ellerbach commented on devices/Ld2410/GateConfiguration.cs:
```diff
@@ -0,0 +1,101 @@
+        /// </summary>
+        /// <remarks>There is no defined unit of measurement for this and should be treated as a percentage. It is recommended to experiment with the values to find the right threshold for the use-case at hand.</remarks>
+        /// <exception cref="ArgumentOutOfRangeException">This value cannot be less than 0 or greater than 100.</exception>
+        public ushort MotionSensitivity
```
if that's a ratio between 0 and 100, why not use the UnitsNet Ratio?

Ellerbach commented on devices/Ld2410/GateConfiguration.cs:
```diff
@@ -0,0 +1,101 @@
+        /// <remarks>There is no defined unit of measurement for this and should be treated as a percentage. It is recommended to experiment with the values to find the right threshold for the use-case at hand.</remarks>
+        /// <exception cref="InvalidOperationException">Cannot set this value for Gate 0 and 1.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">This value cannot be less than 0 or greater than 100.</exception>
+        public ushort RestSensitivity
```
same here with Ratio

Ellerbach reviewed (commented):
Amazing work, few more minor comments compare to the great quality! We're almost there. thanks again for all this hard work!

Ellerbach reviewed (approved):
One more time: great work! Thanks again for bringing this binding in nanoFramework!
